PyInstaller/building/imphook.py

```diff
@@ -12,18 +12,453 @@
+            (e.g., `datas`) with the `format_binaries_and_datas()` helper.
+        """
+
+        # Note that the passed module graph is already a weak reference,
+        # avoiding circular reference issues. See ModuleHookCache.__init__().
+        self.module_graph = module_graph
+        self.module_name = module_name
+        self.hook_filename = hook_filename
+
+        # Name of the in-memory module fabricated to refer to this hook script.
+        self.hook_module_name = (
+            hook_module_name_prefix + self.module_name.replace('.', '_'))
+
+        # Attributes defined by lazily loading this hook.
+        self._scope = None
+
+    ## Properties
+
+    @property
+    def scope(self):
+        """
+        In-memory module of this hook script's interpreted contents, lazily
+        loaded on the first access of this property.
+        """
+
+        # If this is the first access of this property, load this hook script.
+        if self._scope is None:
+            logger.info(
+                'Loading module hook "%s"...',
+                os.path.basename(self.hook_filename))
+            self._scope = importlib_load_source(
+                self.hook_module_name, self.hook_filename)
+
+            # Sanitize all top-level attributes defined by this hook script.
+            self._sanitize_hook_attrs()
+
+        # Return the previously loaded in-memory module.
+        return self._scope
+
+
+    def _sanitize_hook_attrs(self):
+        """
+        Sanitize all top-level attributes defined by this hook script.
+
+        This method converts both the `binaries` and `datas` hook attributes
+        from lists of hook-style 2-tuples into sets of `TOC`-style 2-tuples, as
+        well as defaulting undefined hook attributes to empty lists or sets.
+
+        See Also
+        ----------
+        `format_binaries_and_datas()`
+            Further details on hook- to `TOC`-style conversions.
+        """
+
+        # Convert lists of hook-style 2-tuples into sets of TOC-style 2-tuples,
+        # defaulting undefined lists to empty sets. This must be done before the
+        # subsequently called _process_hook_func() method appends already
+        # converted TOC-style 2-tuples to these sets.
+        for attr_name in ('binaries', 'datas'):
+            setattr(    self.scope, attr_name, set(format_binaries_and_datas(
+                getattr(self.scope, attr_name, list()))))
+
+        # Convert list attributes in which order is insignificant to sets,
+        # defaulting undefined lists to empty sets.
+        for attr_name in ('excludedimports',):
+            setattr(    self.scope, attr_name, set(
+                getattr(self.scope, attr_name, list())))
+
+        # Default all remaining undefined lists to empty. Since order is
+        # significant in these lists, these lists must not be converted to sets.
+        for attr_name in ('hiddenimports',):
+            if  getattr(self.scope, attr_name, None) is None:
```

Comment:
Because laziness and ineptitude. They are [virtues](http://blog.teamtreehouse.com/the-programmers-virtues), I tell you!

<sup>...I'll fix this.</sup>